include/dynamatic/Analysis/ControlDependenceAnalysis.h

```diff
@@ -0,0 +1,125 @@
+//===- ControlDependenceAnalysis.h - Control dependence analyis utilities
+//----------*- C++ -*-===//
+//
+// Dynamatic is under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// Declares some utility functions useful to analyzing the control dependencies
+// between basic blocks of the CFG.
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef DYNAMATIC_ANALYSIS_CONTROLDEPENDENCEANALYSIS_H
+#define DYNAMATIC_ANALYSIS_CONTROLDEPENDENCEANALYSIS_H
+
+#include "mlir/Analysis/CFGLoopInfo.h"
+#include "mlir/Dialect/Func/IR/FuncOps.h"
+#include "mlir/IR/Dominance.h"
+#include "mlir/Pass/AnalysisManager.h"
+
+namespace dynamatic {
+
+/// Analysis to maintain the control dependencies of basic blocks throughout the
+/// IR's lifetime. Query at the beginning of Dynamatic passes using
+/// `getAnalysis<ControlDependenceAnalysis>()` and cache it to avoid
+/// recomputations for further passes using
+/// `markAnalysesPreserved<ControlDependenceAnalysis>()` at the end.
+///
+class ControlDependenceAnalysis {
+public:
+  /// Constructor called automatically by
+  /// `getAnalysis<ControlDependenceAnalysis>()` if the analysis is not already
+  /// cached. It expects to be passed a FuncOp operation where it can loop over
+  /// its blocks (mimicking the BBs of the CFG)
+  ControlDependenceAnalysis(mlir::Operation *operation) {
+    // type-cast it into mlir::func::FuncOp
+    mlir::func::FuncOp funcOp = mlir::dyn_cast<mlir::func::FuncOp>(operation);
+    if (funcOp) {
+      identifyAllControlDeps(funcOp);
+      identifyForwardControlDeps(funcOp);
+    } else {
+      // report an error indicating that the anaylsis is instantiated over an
+      // inappropriate operation
+      llvm::errs() << "ControlDependenceAnalysis is instantiated over an "
+                      "operation that is not FuncOp!\n";
+    }
+  };
+
+  // return all BBs that the block in the argument is control dependent on
+  void returnAllControlDeps(
+      mlir::Block *block,
+      llvm::SmallVector<mlir::Block *, 4> &returned_all_control_deps);
```

Comment:
- It looks a bit strange that this function is called "returnAllControlDeps" but it is void...
- camelCase here (and everywhere else for your variables/arguments): returnedAllControlDeps